container_pipeline/workers/delivery.py

```diff
@@ -0,0 +1,102 @@
+        super(DeliveryWorker, self).__init__(logger, sub, pub)
+        self.openshift = Openshift(logger=self.logger)
+
+    def handle_job(self, job):
+        """Handles a job meant for delivery worker"""
+        # TODO: this needs to be addressed after addressing CentOS#278
+        self.logger.info('Starting delivery for job: {}'.format(job))
+
+        success = self.deliver_build(job)
+
+        if success:
+            self.handle_delivery_success(job)
+        else:
+            self.handle_delivery_failure(job)
+
+    def deliver_build(self, job):
+        """
+        Runs an `oc build` with the `run_delivery.sh` script as a part of build
+        template. It mainly changes the tag of the image from a test tag
+        generated by build process to the tag desired by user as mentioned in
+        `desired-tag` field in cccp.yml
+        """
+        namespace = get_job_name(job)
+        project = hashlib.sha224(namespace).hexdigest()
+
+        try:
+            self.openshift.login()
+            # start the 'delivery' build
+            delivery_id = self.openshift.build(project, 'delivery')
+            if not delivery_id:
+                return False
+        except OpenshiftError as e:
+            logger.error(e)
+            return False
+        finally:
+            self.openshift.delete_pods(project, delivery_id)
+
+        delivery_status = self.openshift.wait_for_build_status(
+            project, delivery_id, 'Complete')
+        logs = self.openshift.get_build_logs(project, delivery_id)
+        delivery_logs_file = os.path.join(job['logs_dir'], 'delivery_logs.txt')
+        self.export_build_logs(logs, delivery_logs_file)
+        return delivery_status
+
+    def handle_delivery_success(self, job):
+        """
+        Sends job details to RPM tracking piece and deletes the job from the
+        tube
+        """
+        self.logger.info('Putting job details to master-tube for tracker\'s'
+                         ' consumption')
+        job['action'] = 'tracking'
+        self.queue.put(json.dumps(job), 'master-tube')
+        job.delete()
+
+    def handle_delivery_failure(self, job):
+        """
+        Puts the job back to the delivery tube for later attempt at delivery
+        and requests to notify the user about failure to deliver
+        """
+        self.queue.put(json.dumps(job))
+        self.logger.info(
+            "Build is not successful putting it to failed build tube")
+        data = {
+            'action': 'notify_user',
+            'namespace': get_job_name(job),
+            'build_status': False,
+            'notify_email': job['notify_email'],
+            'build_logs_file': os.path.join(
+                job['logs_dir'], 'build_logs.txt'),
+            'project_name': get_project_name(job),
+            'job_name': job['jobid'],
+            'TEST_TAG': job['TEST_TAG']}
+        self.logger.debug('Notify build failure: {}'.format(data))
+        self.notify(data)
```

Comment:
@bamachrn I have copied this piece from build worker and my understanding is that `self.queue.put(json.dumps(job))` puts the job back in the queue so that it can be picked up later. But then what are retries for? I'm sure @rtnpro can answer this better.

`self.notify(data)` does the job of notifying the user of the failure in doing delivery.